### PR TITLE
Cuda multi-GPU support: Make some variables device-specific, update Kokkos::fence

### PR DESCRIPTION
**Pull request by masterleinad**
Part of #6091. Some of the `static` `CudaInternal` variables should be created per device as the comments in the code are already saying. Namely, these are `constantMemHostStaging`, `constantMemReusable`, and `constantMemMutex`.
This pull request turns each of them into a `static std::map` that is lazily initialized. To properly deallocate the respective memory in `impl_finalize()`, we also have to store the devices that `Kokkos` is using inside another lazily initialized `std::set`.
This allows us to also update `cuda_device_synchronize` resp. `Kokkos::fence` already to fence all devices that we know of.

I decided to `fence` all devices in `cuda_device_synchronize` instead of adding a `device id` parameter and run through all of them at the call site since it appeared that this matches the implied behavior the best and requires fewer changes.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please

**Comment by masterleinad** [comment]
The failures in two Cuda build are appear to be unrelated to this pull request and more of a machine issue.

File: core/src/Cuda/Kokkos_Cuda_Instance.cpp [hunk 1]
```diff
@@ -313,6 +313,17 @@ void CudaInternal::initialize(cudaStream_t stream) {
...
+  // Allocate a staging buffer for constant mem in pinned host memory
+  // and an event to avoid overwriting driver for previous kernel launches
+  if (!constantMemHostStagingPerDevice[m_cudaDev])
+    KOKKOS_IMPL_CUDA_SAFE_CALL((cuda_malloc_host_wrapper(
+        reinterpret_cast<void **>(&constantMemHostStagingPerDevice[m_cudaDev]),
+        CudaTraits::ConstantMemoryUsage)));
+
+  if (!constantMemReusablePerDevice[m_cudaDev])
+    KOKKOS_IMPL_CUDA_SAFE_CALL(
+        (cuda_event_create_wrapper(&constantMemReusablePerDevice[m_cudaDev])));
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Instance.cpp:325]: Move this block you are adding right after the `CudaInternal::cuda_devices.insert(m_cudaDev);`

File: core/src/Cuda/Kokkos_Cuda_Instance.cpp [hunk 2]
```diff
@@ -615,27 +626,21 @@ Kokkos::Cuda::initialize WARNING: Cuda is allocating into UVMSpace by default
...
-  KOKKOS_IMPL_CUDA_SAFE_CALL(
-      cudaFreeHost(Impl::CudaInternal::constantMemHostStaging));
+  for (auto &cuda_device : Kokkos::Impl::CudaInternal::cuda_devices) {
+    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaSetDevice(cuda_device));
+    KOKKOS_IMPL_CUDA_SAFE_CALL(
+        cudaFreeHost(Kokkos::Impl::CudaInternal::constantMemHostStagingPerDevice
+                         [cuda_device]));
+    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaEventDestroy(
+        Kokkos::Impl::CudaInternal::constantMemReusablePerDevice[cuda_device]));
+  }
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_Instance.cpp:643]: I don't think it matters.

File: core/src/Cuda/Kokkos_Cuda_Instance.hpp [hunk 3]
```diff
@@ -116,11 +120,11 @@ class CudaInternal {
...
-  // FIXME_CUDA: these want to be per-device, not per-stream...  use of 'static'
-  //  here will break once there are multiple devices though
-  inline static unsigned long* constantMemHostStaging = nullptr;
-  inline static cudaEvent_t constantMemReusable       = nullptr;
-  inline static std::mutex constantMemMutex;
+  inline static std::set<int> cuda_devices = {};
+  inline static std::map<int, unsigned long*> constantMemHostStagingPerDevice =
+      {};
+  inline static std::map<int, cudaEvent_t> constantMemReusablePerDevice = {};
+  inline static std::map<int, std::mutex> constantMemMutexPerDevice;
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_Instance.hpp:127]: No, I assume these are holding few elements and would never be called in a performance-critical code path.
If we cared about performance, we would rather use a `std::vector`(or `small_vector`) here anyway and manually search through it where necessary.

File: core/src/Cuda/Kokkos_Cuda_Instance.cpp (hunk 1 above)
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_Instance.cpp]: Sure.

File: core/src/Cuda/Kokkos_Cuda_Instance.cpp [hunk 4]
```diff
@@ -615,27 +626,21 @@ Kokkos::Cuda::initialize WARNING: Cuda is allocating into UVMSpace by default
...
 void Cuda::impl_finalize() {
   (void)Impl::cuda_global_unique_token_locks(true);
-
   desul::Impl::finalize_lock_arrays();  // FIXME
 
-  KOKKOS_IMPL_CUDA_SAFE_CALL(
-      cudaEventDestroy(Impl::CudaInternal::constantMemReusable));
-  KOKKOS_IMPL_CUDA_SAFE_CALL(
-      cudaFreeHost(Impl::CudaInternal::constantMemHostStaging));
+  for (auto &cuda_device : Kokkos::Impl::CudaInternal::cuda_devices) {
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/Cuda/Kokkos_Cuda_Instance.cpp]: Just a habit of writing range-based for loops with references.

File: core/src/Cuda/Kokkos_Cuda_Instance.hpp (hunk 3 above)
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Instance.hpp:127]: Consistency
```suggestion
  inline static std::map<int, std::mutex> constantMemMutexPerDevice = {};
```

**Review by dalg24 — APPROVED** [review]
(with the `= {}` change for consistency)

**Review by tcclevenger — APPROVED** [review]

**Comment by dalg24** [comment]
Failures are unrelated.  I intend to merge when/if the last CUDA build passes

* massive runtime failures in HIP ROCm 5.6 build that are clearly unrelated
* machine issue in gcc build
```
[ 28%] Building CXX object core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_Serial1.dir/serial/TestSerial_HostSharedPtr.cpp.o
Cannot contact CpuNode6: hudson.remoting.ChannelClosedException: Channel "hudson.remoting.Channel@12972753:JNLP4-connect connection from 10.64.193.57/10.64.193.57:34306": Remote call on JNLP4-connect connection from 10.64.193.57/10.64.193.57:34306 failed. The channel is closing down or has closed down
wrapper script does not seem to be touching the log file in /var/jenkins/workspace/Kokkos_PR-6753@tmp/durable-05336f30
(JENKINS-48300: if on an extremely laggy filesystem, consider -Dorg.jenkinsci.plugins.durabletask.BourneShellScript.HEARTBEAT_CHECK_INTERVAL=86400)
```
* in openmptarget build
```
49: [ RUN      ] simd.device_reductions
49: "PluginInterface" error: Failure to synchronize stream (nil): Error in cuStreamSynchronize: an illegal memory access was encountered
49: Libomptarget error: Consult https://openmp.llvm.org/design/Runtimes.html for debugging options.
49: Kokkos_OpenMPTarget_ParallelFor_Range.hpp:54:1: Libomptarget fatal error 1: failure of target construct while offloading is mandatory
49/49 Test #49: Kokkos_UnitTest_SIMD ........................Subprocess aborted***Exception:   1.27 sec
```

**Pull request merged**